modules/runner/src/integrationTest/java/org/apache/ignite/internal/configuration/storage/ItRebalanceDistributedTest.java

```diff
@@ -534,13 +555,121 @@ void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws
         checkInvokeDestroyedPartitionStorages(newNode, TABLE_NAME, 0);
     }
 
+    @Test
+    void testRebalanceWithTheSameNodes() throws Exception {
+        Node node = getNode(0);
+
+        createZone(node, ZONE_NAME, 1, 1);
+
+        clearSpyInvocations();
+
+        createTable(node, ZONE_NAME, TABLE_NAME);
+
+        assertTrue(waitForCondition(() -> getPartitionClusterNodes(node, 0).size() == 1, AWAIT_TIMEOUT_MILLIS));
+
+        alterZone(node, ZONE_NAME, 3);
+
+        checkAssignments();
+
+        updateAssignments();
+
+        checkAssignments();
+
+        verifyThatRaftNodesAndReplicasWereStartedOnlyOnce();
+    }
+
+    private void clearSpyInvocations() {
+        clearInvocations(getNode(0).getRaftManager());
+        clearInvocations(getNode(0).getReplicaManager());
+        clearInvocations(getNode(1).getRaftManager());
+        clearInvocations(getNode(1).getReplicaManager());
+        clearInvocations(getNode(2).getRaftManager());
+        clearInvocations(getNode(2).getReplicaManager());
+    }
+
+    private void checkAssignments() throws Exception {
+        waitPartitionPlannedAssignmentsSyncedToExpected(0, 0);
+        waitPartitionPendingAssignmentsSyncedToExpected(0, 0);
+        waitPartitionAssignmentsSyncedToExpected(0, 3);
+    }
+
+    /**
+     * Update pending and planned assignments to start a rebalance.
+     *
+     * @throws Exception If fail.
+     */
+    private void updateAssignments() throws Exception {
+        Collection<String> dataNodes = new HashSet<>();
+        dataNodes.add(getNode(0).getName());
+        dataNodes.add(getNode(1).getName());
+        dataNodes.add(getNode(2).getName());
+
+
+        Set<Assignment> pendingAssignments = AffinityUtils.calculateAssignmentForPartition(dataNodes, 0, 2);
+        Set<Assignment> plannedAssignments = AffinityUtils.calculateAssignmentForPartition(dataNodes, 0, 3);
+
+        byte[] bytesPendingAssignments = ByteUtils.toBytes(pendingAssignments);
+        byte[] bytesPlannedAssignments = ByteUtils.toBytes(plannedAssignments);
+
+        TablePartitionId partId = new TablePartitionId(1, 0);
+
+        ByteArray partAssignmentsStableKey = stablePartAssignmentsKey(partId);
+        ByteArray partAssignmentsPendingKey = pendingPartAssignmentsKey(partId);
+        ByteArray partAssignmentsPlannedKey = plannedPartAssignmentsKey(partId);
+
+        SimpleCondition condition = exists(partAssignmentsStableKey);
```

Comment:
I think you can use simple `MetaStorageManager#putAll`

Reply:
Fixed.